tests/components/rest/test_sensor.py

```diff
@@ -1,696 +1,665 @@
+            "sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "GET",
+                "json_attributes": ["key"],
+                "name": "foo",
+                "unit_of_measurement": DATA_MEGABYTES,
+                "verify_ssl": "true",
+                "timeout": 30,
+                "headers": {"Accept": "text/xml"},
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+    state = hass.states.get("sensor.foo")
+    assert state.state == '{ "key": "some_json_value" }'
+
+
+@respx.mock
+async def test_update_with_json_attrs_no_data(hass, caplog):
+    """Test attributes when no JSON result fetched."""
+
+    respx.get(
+        "http://localhost",
+        status_code=200,
+        headers={"content-type": CONTENT_TYPE_JSON},
+        content="",
+    )
+    assert await async_setup_component(
+        hass,
+        "sensor",
+        {
+            "sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "GET",
+                "value_template": "{{ value_json.key }}",
+                "json_attributes": ["key"],
+                "name": "foo",
+                "unit_of_measurement": DATA_MEGABYTES,
+                "verify_ssl": "true",
+                "timeout": 30,
+                "headers": {"Accept": "text/xml"},
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+    state = hass.states.get("sensor.foo")
+    assert state.state == STATE_UNKNOWN
+    assert state.attributes == {"unit_of_measurement": "MB", "friendly_name": "foo"}
+    assert "Empty reply" in caplog.text
+
+
+@respx.mock
+async def test_update_with_json_attrs_not_dict(hass, caplog):
+    """Test attributes get extracted from a JSON result."""
+
+    respx.get(
+        "http://localhost",
+        status_code=200,
+        headers={"content-type": CONTENT_TYPE_JSON},
+        content='["list", "of", "things"]',
+    )
+    assert await async_setup_component(
+        hass,
+        "sensor",
+        {
+            "sensor": {
+                "platform": "rest",
+                "resource": "http://localhost",
+                "method": "GET",
+                "value_template": "{{ value_json.key }}",
+                "json_attributes": ["key"],
+                "name": "foo",
+                "unit_of_measurement": DATA_MEGABYTES,
+                "verify_ssl": "true",
+                "timeout": 30,
+                "headers": {"Accept": "text/xml"},
+            }
+        },
+    )
+    await hass.async_block_till_done()
+    assert len(hass.states.async_all()) == 1
+
+    state = hass.states.get("sensor.foo")
+    assert state.state == ""
+    assert state.attributes == {"unit_of_measurement": "MB", "friendly_name": "foo"}
+    assert "not a dictionary or list" in caplog.text
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)